### PR TITLE
adapter: Give mz_introspection USAGE on clusters

### PR DESCRIPTION
**Pull request by jkosh44**
This commit gives the mz_introspection role USAGE privileges on all clusters. This is required so that the prometheus-exporter can get information on cluster replicas by executing replica targeted queries.

Fixes https://github.com/MaterializeInc/cloud/issues/6709

### Motivation
This PR fixes a recognized bug.

### Checklist

- [X] This PR has adequate test coverage / QA involvement has been duly considered.
- [X] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [X] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [X] If this PR will require changes to cloud orchestration, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [X] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - There are no user-facing behavior changes.

### CODE REVIEW

File: src/stash/src/upgrade/v22_to_v23.rs [hunk 1]
```diff
@@ -0,0 +1,457 @@
...
+                let new_key = key.clone().into();
+                let new_value = value.clone().into();
+                updates.push(MigrationAction::Update(key.clone(), (new_key, new_value)));
+            }
+
+            updates
+        })
+        .await?;
+
+    CLUSTER_COLLECTION
```
Comment (ParkMyCar) [inline src/stash/src/upgrade/v22_to_v23.rs:65]: Should we be updating the `CLUSTER_COLLECTION` in the Stash initialization too?

**Review by ParkMyCar — COMMENTED** [review]
Just one comment about the Stash initialization flow, otherwise this looks good!

File: src/stash/src/upgrade/v22_to_v23.rs (hunk 1 above)
Comment (jkosh44) [reply src/stash/src/upgrade/v22_to_v23.rs:65]: Yes, good catch. Just pushed an update.

File: src/stash/src/upgrade/v22_to_v23.rs [hunk 2]
```diff
@@ -0,0 +1,457 @@
...
+pub async fn upgrade(tx: &'_ mut Transaction<'_>) -> Result<(), StashError> {
+    const DEFAULT_PRIVILEGES_COLLECTION: TypedCollection<
+        objects_v22::DefaultPrivilegesKey,
+        objects_v22::DefaultPrivilegesValue,
+    > = TypedCollection::new("default_privileges");
+    const CLUSTER_COLLECTION: TypedCollection<objects_v22::ClusterKey, objects_v22::ClusterValue> =
+        TypedCollection::new("compute_instance");
+
+    DEFAULT_PRIVILEGES_COLLECTION
+        .migrate_to(tx, |entries| {
```
[the remark this answers is not in the document]
Comment (jkosh44) [reply src/stash/src/upgrade/v22_to_v23.rs:41]: > This pattern of inserting some new object, without migrating the existing ones, seems like something common, but right now it's quite verbose.

I was thinking the exact same thing.

> Do you think having a helper like `insert_new(...)` that doesn't require you to implement from `vX` for `vY` would be helpful?

Yes, I think that would be very helpful. Hopefully we can design it in such a way that we don't need to take a new snapshot of `objects.proto` if the contents aren't changing.

**Review by ParkMyCar — APPROVED** [review]

File: src/stash/src/upgrade/v22_to_v23.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (ParkMyCar) [reply src/stash/src/upgrade/v22_to_v23.rs:41]: Great! I'm not sure if we're able to skip snapshoting `objects.proto`, but I have an idea on how to make this process much easier, I filed https://github.com/MaterializeInc/materialize/issues/20163 to track it

**Pull request merged**